lib/models.dart

```diff
@@ -36,27 +48,38 @@ class BikeState with _$BikeState {
       BikeRegion? region,
       @Default(false) bool modeLock,
       @Default(false) bool selected,
-      @Default(0) int color}) = _BikeState;
+      @Default(0) double speedKM,
+      @Default(0) double speedMI,
+      @Default(0) int range,
+      @Default(0) double battery,
+      @Default(0) double voltage,
+      @Default(0) int color,
+      @Default('metric') String speedMetric,
+      @Default('percent') String batteryMetric}) = _BikeState;
 
   factory BikeState.fromJson(Map<String, Object?> json) =>
       _$BikeStateFromJson(json);
 
   factory BikeState.defaultState(String id) {
     return BikeState(
-        id: id, mode: 0, light: false, assist: 0, name: getName(seed: id));
+        id: id, mode: 0, light: false, assist: 0, name: getName(seed: id), speedKM: 0.0, speedMI: 0.0, range: 0, battery: 0.0, voltage: 0.0);
   }
 
-  BikeState updateFromData(List<int> data) {
-    const lightIdx = 4;
-    const modeIdx = 5;
-    const assistIdx = 2;
-    final region = _guessRegion(data[modeIdx]);
-    final newmode = _modeFromRead(data[modeIdx]);
+  BikeState updateFromData(List<int> SETTINGS, List<int> RIDE, List<int> MOTION) {
+    //SETTINGS	0x03	0x00
+    //RIDE	    0x02	0x03
+    //MOTION	  0x02  0x01
     return copyWith(
-        light: data[lightIdx] == 1,
-        mode: newmode,
-        assist: data[assistIdx],
-        region: region);
+      light: SETTINGS[4] == 1,
+      mode: _modeFromRead(SETTINGS[5]),
+      assist: SETTINGS[2],
+      region: _guessRegion(SETTINGS[5]),
+      speedKM: _speedKM(MOTION),
+      speedMI: _speedMI(MOTION),
```

Comment:
I guess I would have flipped what this is doing, where I'd make a speedRaw state that is just the `MOTION` value, then put getters on the state to convert that into km or mi. Same with battery percentage and voltage. It's just 1/2 the state to manage.

Reply:
Just confirming that I have it right.

so instead of `speedKM` & `speedMI `
have it `motion: MOTION`

then on bike.dart convert it by doing `bike._speedKM(bike.motion);`
like likewise for the rest of the variables?

Reply:
Kind of like `viewMode`, where it takes the raw mode value (from 0 to 3) and adds 1 (from 1 to 4):

```
  String get viewMode {
    return (mode + 1).toString();
  }
```

Then the app just calls `bike.viewMode` to get the value to show users.

State could do somthing like that:

```
  String get viewSpeed {
    if (speedMetric == 'm'){
       return '$_speedKM(speedRaw) km/h' ;
  ... etc
  }
```

Then the app just calls `bike.viewSpeed` in the widget and the right conversion and units come out.

Reply:
Also, it would be nice to guess the units of measure based on the guessed region if the user hasn't selected values yet (EU is metric, US is imperial). In that vein, the options for unit selection are not type safe like region is. Ideally they would be Enums like BikeRegion.

Reply:
> Also, it would be nice to guess the units of measure based on the guessed region if the user hasn't selected values yet (EU is metric, US is imperial). In that vein, the options for unit selection are not type safe like region is. Ideally they would be Enums like BikeRegion.

I originally thought of that, However since I ride using region US (for stronger torque) in an Metric mindset. I scrapped it.

I guess I can add it, since its all being saved in db, I can default select based of region,

Reply:
Right, it's just guessing the default so _most_ users will get the right value. You will too, since switching the region to US will happen _after_ the guess which would have already guessed KM based on your factory region of EU. Once the guess happens the setting is saved and will only be guessed again if the phones data is wiped.